### DISCUSSION TITLE
Links to local file system

### DISCUSSION THREAD

**Discussion by cplerch** [TriliumNext/Trilium#1549] [Ideas]
How about links to the local file system by using to file:// url protocol?
Well, right now it just doesn't work: nothing happens when you click on such link.
Anybody out there who would find this feature useful too?
Please comment, thanks.

**Comment by zadam** [comment]
Hi, this was already brought up but is not supported yet: https://github.com/zadam/trilium/issues/1293 and https://github.com/zadam/trilium/issues/965

The biggest issue is support in CKEditor: https://github.com/ckeditor/ckeditor5/issues/8248

**Reply by eduardorubim** [reply]
this has been fixed
https://github.com/ckeditor/ckeditor5/issues/14304

**Comment by cplerch** [comment]
Yes, but:
LukaszGudel <https://github.com/LukaszGudel> commented on 16 Dec 2020
<https://github.com/ckeditor/ckeditor5/issues/8686#issuecomment-746130680>
We are aware of this issue and you can read more about it here - #8248
<https://github.com/ckeditor/ckeditor5/issues/8248>. Also, feel free to add
👍 to the first post and join the discussion to bump its popularity!

Currently "unknown" protocol in a link is replaced by '#'. But only inside
the editor view. If you use editor.getData() then in returned string you
will find <a> tag with unchanged href attribute.

(See: https://github.com/ckeditor/ckeditor5/issues/8686)


What does this mean for the CKEditor instance as configured and used by
trilium-notes?

Am Di., 19. Jan. 2021 um 13:56 Uhr schrieb zadam <notifications@github.com>:

**Reply by zadam** [reply]
It means that it's not possible to open local filesystem link from Trilium.

**Reply by ghost** [reply]
Hi there,

but we have that **_open file externally_** command ...
Till yet, I can drag&drop a file, let say an .odt (libre writer).
Sure, trilium says, that it can't display this note, but click'n on Open opens it with the default app (libreoffice in this case),
So, when I can apply the open-command to a file link ?

Maybe that's a possibility for short implementation.
Kai

**Comment by laundmo** [comment]
I made a trilium script to make file paths in normal text double-clickable.

It bypasses all of CKEditor and just registers a global double click handler which checks whether the clicked element matches a file path regex and is italicized. That means it doesn't show a different cursor or any indication of the file being opened. It will open folders with the explorer and files with their default application.

https://gist.github.com/laundmo/de1f386fac9f9e797fd77022d63967c9

**Comment by connectluole** [comment]
@laundmo , very thanks for the link file open script, and based on your script, i just hack a little , which suit the
trilium default open link operation that is ctrl+one_click to open a link externally, and i just hack the script to 
make the **"ctrl+double_click" to open the local file .**
Also, the new script no longer require the italicized, and not require the path alone in one line.
**it can fit many format** , once you type ctrl + double click on this line , it can auto open the path by your system's default software.
Demo is given below:
![show](https://github.com/zadam/trilium/assets/16472308/54f2a775-03ee-43dc-b6e2-1c532be8d922)

---
code is given below

```
/* readme 
create one javascript front type code note in your trilium which contain 
below javascript code  , and add the    #widget    attribute to it .
*/
async function OpenLocalPath(event) {
    // Check if the event was a Ctrl+double click.
    if (event.ctrlKey) {
        var path = event.target.innerText;
        console.log(`>> ctrl + double_click get line content:\n${path}`)
        //------ Check if the text content contain a file path.   ------
        if (/[A-Z]:\\.*/.test(path)) {
            var path = path.match(/[A-Z]:\\.*/)[0]
            console.log(`>> Open local path: ${path}`)
            // Use the PathLinkerApi to open the file at the path.
            await document.PathLinkerApi.runAsyncOnBackendWithManualTransactionHandling(async (path) => { 
                const shell = require('electron').shell;
                await shell.openPath(path);
                return;
            }, [path]);
        }
    }
}

const TEMPLATE = `
    <div style="padding: 10px; border-top: 1px solid var(--main-border-color); contain: none;">
    <script>
    ${OpenLocalPath.toString()}
    document.addEventListener('dblclick', OpenLocalPath);
    </script>
    </div>`;

class PathLinkerWidget extends api.NoteContextAwareWidget {
    constructor(...args) {
        super(...args);
        this.balloonEditorCreate = null;
    }
    
    get position() { 
        // higher value means position towards the bottom/right
        return 100; 
    } 

    get parentWidget() { return 'center-pane'; }

    doRender() {
        this.$widget = $(TEMPLATE);
        this.$widget.hide();
        
        // store API in document to access it from callback
        document.PathLinkerApi = api;
        return this.$widget;
    }
}

let widget = new PathLinkerWidget();
console.log(">> Loaded PathLinkerWidget");
module.exports = widget;
```

it will get below operation

**Reply by MarquezRomulo** [reply]
Hi all,

How can I install this script inside Trilium? I'm not Trilium expert sorry!

Thank you!

**Reply by ocean-haiyang** [reply]
I also want to know. Im on linux

**Reply by ocean-haiyang** [reply]
> Hi all,
> 
> How can I install this script inside Trilium? I'm not Trilium expert sorry!
> 
> Thank you!

I figured it out. 

1. Make a new note
2. add #widget attribute
3. make it a jsfrontend code note
4. add script
5. restart trilium

**Comment by connectluole** [comment]
By the way , @zadam , can you consider to give a default operation that treate the local file as the common external link ?

**Comment by grosserfrank** [comment]
Hi, is it possible to a portable path like **{trilium_folder}\files\filexy.pdf**
???

**Comment by grosserfrank** [comment]
e.G. .\files\filexy.pdf

**Reply by andjar** [reply]
@grosserfrank this modified function will look for "./" instead, but you have to manually set `const basePath = 'C:\\Users\\you\\trilium\\assets';` to the right base folder.

For example, `./2024-05-27/hi.pdf` will now point to `C:/Users/you/trilium/assets/2024-05-27/hi.pdf`

```js
async function OpenLocalPath(event) {
    // Check if the event was a Ctrl+double click.
    if (event.ctrlKey) {
        var path = event.target.innerText;
        console.log(`>> ctrl + double_click get line content:\n${path}`)
        //------ Check if the text content contain a file path.   ------
        if (path.startsWith('./')) {
            const basePath = 'C:\\Users\\you\\trilium\\assets';
            // Remove the './' from the start of the path
            let relativePath = path.substring(2);
            // Combine the base path with the relative path
            let fullPath = `${basePath}\\${relativePath}`;
            console.log(`>> Open local path: ${fullPath}`)
            // Use the PathLinkerApi to open the file at the path.
            await document.PathLinkerApi.runAsyncOnBackendWithManualTransactionHandling(async (fullPath) => { 
                const shell = require('electron').shell;
                await shell.openPath(fullPath);
                return;
            }, [fullPath]);
        }
    }
}
```

**Comment by CobriMediaJulien** [comment]
Is this portable to the new trilium next version? I tried it but it doesn´t work because require("electron") is not defined. I tried to fix the script but with my knowledge i failed. Does sb has a hint for me how to adapt the skripts for trilium next?

**Reply by connectluole** [reply]
this script is only available for the windows desktop trilium app.

**Reply by CobriMediaJulien** [reply]
allright, ty for responding. for information: trilium is in maintance mode. We do continue the development process here:

https://github.com/orgs/TriliumNext/discussions

**Reply by CobriMediaJulien** [reply]
Updated scripts for trilium next

**Comment by ocean-haiyang** [comment]
I made a script for linux https://github.com/zadam/trilium/discussions/4907

**Comment by movingsubject** [comment]
It would be great having this officially implemented

**Reply by ocean-haiyang** [reply]
submitted :) https://github.com/TriliumNext/Notes/pull/1141
